### PR TITLE
Add index on school_infos for country, school_name, and school_type

### PR DESCRIPTION
**Pull request by drizco**
a query that checks for existing `school_infos` entries was impacting performance on the production db. the slow query was searching for an existing record matching the `country` and `school_name` columns. at the moment, a school_info record will either have a `school_id` (nces school) which already has an index, or it will have a `country` and either a `school_name` or a `school_type`. so the thought was to add an index on those three columns (`country, `school_name` and `school_type`). after creating a production db clone and connecting to it, I was able to analyze the query with and without an index.

BEFORE ADDING INDEX:
```
-> Limit: 1 row(s)  (cost=2763.80 rows=0.5) (actual time=2784.514..2784.514 rows=0 loops=1)
    -> Filter: ((school_infos.country = 'ID') and (school_infos.state is null) and (school_infos.zip is null) and (school_infos.school_name is null) and (school_infos.validation_type = 'none') and (school_infos.school_type = 'noSchoolSetting') and (school_infos.school_district_other is null) and (school_infos.school_district_name is null) and (school_infos.school_id is null) and (school_infos.school_other is null) and (school_infos.full_address is null))  (cost=2763.80 rows=0.5) (actual time=2784.513..2784.513 rows=0 loops=1)
        -> Index lookup on school_infos using fk_rails_951bceb7e3 (school_district_id=NULL), with index condition: (school_infos.school_district_id is null)  (cost=2763.80 rows=325394) (actual time=0.215..2690.459 rows=841890 loops=1)
```

AFTER ADDING INDEX:
```
-> Limit: 1 row(s)  (cost=0.26 rows=0.05) (actual time=0.021..0.021 rows=0 loops=1)
    -> Filter: ((school_infos.state is null) and (school_infos.zip is null) and (school_infos.validation_type = 'none') and (school_infos.school_district_id is null) and (school_infos.school_district_other is null) and (school_infos.school_district_name is null) and (school_infos.school_id is null) and (school_infos.school_other is null) and (school_infos.full_address is null))  (cost=0.26 rows=0.05) (actual time=0.020..0.020 rows=0 loops=1)
        -> Index lookup on school_infos using idx_school_infos_name_country_type (school_name=NULL, country='ID', school_type='noSchoolSetting'), with index condition: (school_infos.school_name is null)  (cost=0.26 rows=1) (actual time=0.020..0.020 rows=0 loops=1)

```

## Links

[slack thread](https://codedotorg.slack.com/archives/C0T0PNTM3/p1733162798362609)

## Testing story

tested manually on a clone of the production `school_infos` table

## Deployment strategy

adding the index to the cloned production db took a little over 6 seconds

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/db/migrate/20241202234601_add_index_to_school_infos_country_and_school_name_and_school_type.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class AddIndexToSchoolInfosCountryAndSchoolNameAndSchoolType < ActiveRecord::Migration[6.1]
+  def change
+    add_index :school_infos, [:school_name, :country, :school_type]
```
Comment (bethanyaconnor) [inline dashboard/db/migrate/20241202234601_add_index_to_school_infos_country_and_school_name_and_school_type.rb]: Very possible this is just Rails magic but I see multiple indices added in the annotations and schema.rb but only one here?

Comment (drizco) [reply dashboard/db/migrate/20241202234601_add_index_to_school_infos_country_and_school_name_and_school_type.rb]: oh hmm... yeah let me take another look. I had previously run a migration that added an index to just country and school_name, and tried to undo the migration and re-apply it with school_type 😬

Comment (drizco) [reply dashboard/db/migrate/20241202234601_add_index_to_school_infos_country_and_school_name_and_school_type.rb]: okay I did a reset of my local db off of `staging-next` and re-ran the migration, so there's only the one added index 😅

**Review by hannahbergam — APPROVED** [review]
Thank you for doing this! And good catch @bethanyaconnor

**Comment by sureshc** [comment]
Let’s set the base branch to be `staging` because I think we want to release this before Hour of Code.

**Comment by drizco** [comment]
I did an interactive rebase, trying to drop those 8 commits that were in staging-next when I created the branch, but when I force pushed github started freaking out saying it's still processing the change. closing this pr in favor of https://github.com/code-dot-org/code-dot-org/pull/62789

Screenshot 2024-12-03 at 12 57 14 PM

**Pull request closed without merging**